src/net_processing.cpp

```diff
@@ -2777,11 +2787,28 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         std::vector<CAddress> vAddrOk;
         int64_t nNow = GetAdjustedTime();
         int64_t nSince = nNow - 10 * 60;
+
+        // Update/increment addr rate limiting bucket.
+        const auto current_time = GetTime<std::chrono::microseconds>();
+        if (peer->m_addr_token_bucket < MAX_ADDR_TO_SEND) {
+            // Don't increment bucket if it's already full
+            auto time_diff = std::max(current_time - peer->m_addr_token_timestamp, 0us);
+            double increment = std::chrono::duration<double>(time_diff).count() * MAX_ADDR_RATE_PER_SECOND;
+            peer->m_addr_token_bucket = std::min<double>(peer->m_addr_token_bucket + increment, MAX_ADDR_TO_SEND);
+        }
+        peer->m_addr_token_timestamp = current_time;
+
+        Shuffle(vAddr.begin(), vAddr.end(), FastRandomContext());
```

Comment:
nit: this entire snippet can be inside `if (!pfrom.HasPermission(NetPermissionFlags::Addr)) {` to avoid making calculations that are not going to be used later. That may be insignificant wrt performance (the shuffle too?) and the deeper indentation may look ugly, so feel free to ignore.

Reply:
No, if `pfrom.HasPermission(NetPermissionFlags::Addr)` is true, the loop still has to run - just the rate-limiting check is skipped.

Reply:
> this entire snippet can be inside ...
> ... the loop still has to run ...

The loop if after the snippet, it will run. Just to clarify:

```diff
--- i/src/net_processing.cpp
+++ w/src/net_processing.cpp
@@ -2790,24 +2790,28 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
         // Store the new addresses
         std::vector<CAddress> vAddrOk;
         int64_t nNow = GetAdjustedTime();
         int64_t nSince = nNow - 10 * 60;
 
-        // Update/increment addr rate limiting bucket.
-        const auto current_time = GetTime<std::chrono::microseconds>();
-        if (peer->m_addr_token_bucket < MAX_ADDR_PROCESSING_TOKEN_BUCKET) {
-            // Don't increment bucket if it's already full
-            const auto time_diff = std::max(current_time - peer->m_addr_token_timestamp, 0us);
-            const double increment = CountSecondsDouble(time_diff) * MAX_ADDR_RATE_PER_SECOND;
-            peer->m_addr_token_bucket = std::min<double>(peer->m_addr_token_bucket + increment, MAX_ADDR_PROCESSING_TOKEN_BUCKET);
-        }
-        peer->m_addr_token_timestamp = current_time;
-
-        Shuffle(vAddr.begin(), vAddr.end(), FastRandomContext());
         const bool rate_limited = !pfrom.HasPermission(NetPermissionFlags::Addr);
+
+        if (rate_limited) {
+            // Update/increment addr rate limiting bucket.
+            const auto current_time = GetTime<std::chrono::microseconds>();
+            if (peer->m_addr_token_bucket < MAX_ADDR_PROCESSING_TOKEN_BUCKET) {
+                // Don't increment bucket if it's already full
+                const auto time_diff = std::max(current_time - peer->m_addr_token_timestamp, 0us);
+                const double increment = CountSecondsDouble(time_diff) * MAX_ADDR_RATE_PER_SECOND;
+                peer->m_addr_token_bucket = std::min<double>(peer->m_addr_token_bucket + increment, MAX_ADDR_PROCESSING_TOKEN_BUCKET);
+            }
+            peer->m_addr_token_timestamp = current_time;
+
+            Shuffle(vAddr.begin(), vAddr.end(), FastRandomContext());
+        }
+
         for (CAddress& addr : vAddr)
```

Does not change the behavior, but is slightly more performant (and deeply indented). Again, feel free to ignore. This comment is just for clarification, not an insistence :)